NEW API for getting, adding, deleting and/or modifying email templates

JonBendtsen opened the pull request:
# NEW API for email templates
This PR introduces a new API endpoint which can GET, DELETE, PUT and POST email templates inside Dolibarr. GET, PUT and DELETE can be done with both id and label

Commit:
2 get functions by Id and Label for email templates

Commit:
have not made contribution yet to this file

JonBendtsen commented:
JSON from getById, which is identical to getByLabel output
```
{
  "module": "",
  "id": 19,
  "entity": 1,
  "date_creation": "",
  "date_validation": null,
  "date_modification": null,
  "tms": "2025-03-08 10:45:43",
  "date_cloture": null,
  "active": 1,
  "content": "*****************",
  "content_lines": "",
  "defaultfortype": 1,
  "email_from": "",
  "email_to": "",
  "email_tocc": "",
  "email_tobcc": "",
  "enabled": "1",
  "fk_user": 0,
  "label": "Order_Validated",
  "lang": "en_DK",
  "joinfiles": "",
  "position": 1,
  "private": 0,
  "topic": "Registration accepted, please see payment instructions inside",
  "type_template": "order_send",
  "datec": 0
}
```

JonBendtsen commented:
![email_template_get_by_id](https://github.com/user-attachments/assets/60d2872c-2ee5-4ecf-b270-fadd7ca422da)
![email_template_get_by_id_json](https://github.com/user-attachments/assets/ae3377da-397d-4eec-9dfa-5dbabf573c8e)

JonBendtsen commented:
Where does the text in the yellow box come from?

![email_template_get_by_label](https://github.com/user-attachments/assets/bffa5efc-306f-4526-975b-43dbbefc6fbd)

frederic34 commented:
@JonBendtsen have you seen this class ?
https://github.com/Dolibarr/dolibarr/blob/2ccfb2dfb6cc627a2c2ea29d665bee83fded4a62/htdocs/core/class/html.formmail.class.php#L2094-L2101

Commit:
fixing phan, phpstan + some text descriptions

JonBendtsen commented:
> @JonBendtsen have you seen this class ?
>

yes @frederic34 I have, and I do not think it does what I want to do, but I think I want to use it later when I send email using a template. 

That file contains 2 classes, which of the 2 classes do you think I am making a replica of?

```
class FormMail extends Form
class ModelMail extends CommonObject
```

JonBendtsen commented:
> @JonBendtsen have you seen this class ?
> 
> https://github.com/Dolibarr/dolibarr/blob/2ccfb2dfb6cc627a2c2ea29d665bee83fded4a62/htdocs/core/class/html.formmail.class.php#L2094-L2101

okay, I see it now, the `class ModelMail extends CommonObject` looks very similar - though I really wonder why the name is ModelMail ??

Commit:
renamed to better reflect the object name

Commit:
some modifications from the old FormModel configuration. + an APIfetch because the fetch function didn't work for me

Commit:
copied who has made contributions from top of html.formmail.class.php

Commit:
this name is just btter, no need to put a c in front

JonBendtsen commented:
> > @JonBendtsen have you seen this class ?
> > https://github.com/Dolibarr/dolibarr/blob/2ccfb2dfb6cc627a2c2ea29d665bee83fded4a62/htdocs/core/class/html.formmail.class.php#L2094-L2101
> 
> okay, I see it now, the `class ModelMail extends CommonObject` looks very similar - though I really wonder why the name is ModelMail ??

Heh, the file itself says:
```
 * TODO Move this class into a file cemailtemplate.class.php
```
okay :-)

Commit:
using tabs not spaces to indent, the spaces came from a cut'n'copy operation

Commit:
delete the old name for the class

Commit:
fix pre-commit checks

Commit:
fix pre-commit checks

JonBendtsen commented:
seems like only 1 file needs this old object

```
JonSweet16:htdocs jonbendtsen$ grep -irl -e "require.*html.formmail.class.php"
./holiday/card.php
./holiday/card_group.php
./core/class/html.formticket.class.php
./fourn/class/fournisseur.facture.class.php
./compta/facture/class/facture.class.php
./website/index.php
./expensereport/card.php
./admin/mails_templates.php
./comm/action/class/actioncomm.class.php
./comm/mailing/card.php
./product/document.php
JonSweet16:htdocs jonbendtsen$ for PF in $( grep -irl -e "require.*html.formmail.class.php" ); do
> grep ModelMail ${PF}
> done
	$tmpmailtemplate = new ModelMail($db);
JonSweet16:htdocs jonbendtsen$ for PF in $( grep -irl -e "require.*html.formmail.class.php" ); do grep -l ModelMail ${PF}; done
./admin/mails_templates.php
```

Commit:
class definition moved to core/class/cemailtemplate.class.php

Commit:
require core/class/cemailtemplate.class.php for the ModelMail object

Commit:
fix some pre-commit check

Commit:
2 new functions to delete by id label. Splitting out function to check for access rights. Still replacing objectlink text

hregis commented:
@JonBendtsen email template use the entity field !!! Multicompany compatibility !

JonBendtsen commented:
> @JonBendtsen email template use the entity field !!! Multicompany compatibility !

yes? but it gets that from `CommonObject` right? when I extend that? shouldn't it inherit this?

```
	/**
	 * @var int 		The environment ID when using a multicompany module
	 */
	public $entity;
```

JonBendtsen commented:
How do I allow for labels having spaces around it? Why? because of a lot of email templates has that.

`(AnswerCandidature)`

`http://localhost:22180/api/index.php/emailtemplates/label/(AnswerCandidature)`

```
Access refused to 192.168.127.1 by SQL or Script injection protection in main.inc.php:analyseVarsForSqlAndScriptsInjection type=2<br>
Try to go back, fix data of your form and resubmit it. You can contact also your technical support.
<!--
page=/api/index.php/emailtemplates/label/(AnswerCandidature) paramtype=2 paramkey=0 paramvalue=/api/index.php/emailtemplates/label/(AnswerCandidature)
-->
```

Commit:
making hregis happy by enabling support for multientities

JonBendtsen commented:
> @JonBendtsen email template use the entity field !!! Multicompany compatibility !

this is probably what you meant
`$sql .= " AND entity = ".((int) $this->entity);`

Commit:
function delete works with ID, but not label + trying to set the error message depending on id or label used for fetching

Commit:
making pre-commit happy that wanted it in CamelCaps

Commit:
giving better 404 explanations when a template is not found

Commit:
fixing param and return values for the check access rights

hregis commented:
> > @JonBendtsen email template use the entity field !!! Multicompany compatibility !
> 
> this is probably what you meant `$sql .= " AND entity = ".((int) $this->entity);`

pas du tout ! :-)

hregis commented:
@JonBendtsen tu fais beaucoup de dégâts pour une finalité qui est simple !

JonBendtsen commented:
> @JonBendtsen email template use the entity field !!! Multicompany compatibility !

hey @hregis

I do not use Multicompany and still some of the entries I made by hand using the GUI has 1 for entity and a lot of the default templates that comes with Dolibarr has a 0 for entity.

How did that happen? (I've started with 18, then 19, and now my production is on 20, and it is in my develop which is 22.0.0-alpha where I see that the entity is different, using phpmyadmin. 

I can still see both email templates in the GUI?

This is the dolibarr.log segment when I look for a template using the label

sql=SELECT e.rowid, e.entity, e.module, e.type_template, e.lang, e.private, e.fk_user, e.datec, e.tms, e.label, e.position, e.defaultfortype, e.enabled, e.active, e.email_from, e.email_to, e.email_tocc, e.email_tobcc, e.topic, e.joinfiles, e.content, e.content_lines FROM llx_c_email_templates as e WHERE e.label = 'order_creation' AND e.entity = 0

hregis commented:
> @hregis  I don't understand french

Me too

Commit:
thinkIfinallyNailedCamelCaps

hregis commented:
@JonBendtsen @eldy Give me some time, I'll look at this with a clear head

JonBendtsen commented:
> > @JonBendtsen email template use the entity field !!! Multicompany compatibility !
> 
> hey @hregis
> 
> I do not use Multicompany and still some of the entries I made by hand using the GUI has 1 for entity and a lot of the default templates that comes with Dolibarr has a 0 for entity.
> 
> How did that happen? (I've started with 18, then 19, and now my production is on 20, and it is in my develop which is 22.0.0-alpha where I see that the entity is different, using phpmyadmin.
> 
> I can still see both email templates in the GUI?
> 
> This is the dolibarr.log segment when I look for a template using the label
> 
> sql=SELECT e.rowid, e.entity, e.module, e.type_template, e.lang, e.private, e.fk_user, e.datec, e.tms, e.label, e.position, e.defaultfortype, e.enabled, e.active, e.email_from, e.email_to, e.email_tocc, e.email_tobcc, e.topic, e.joinfiles, e.content, e.content_lines FROM llx_c_email_templates as e WHERE e.label = 'order_creation' AND e.entity = 0

I just tried making a new template by hand in the GUI and it got the entity value of 1 ?

hregis commented:
@JonBendtsen just a moment
please

Commit:
forgot the little e. prefix when checking if entity is correct

hregis commented:
@eldy j'ai besoin de réfléchir !

hregis commented:
@eldy je n'arrive plus à réfléchir en temps réel ! lol

Commit:
noting that enabled usage is unknown

Commit:
adding index function that finds email templates

Commit:
unset some fields email templates does not have + indenting the index docstring

Commit:
adding accesstype to checkAccessRight function, defining required fields and integer fields + not unsetting datec because it is in the database

Commit:
function to create a new email template + a function to test if we already have an object with the same label

Commit:
adding forgotten db->escape function to apifetch using a label

Commit:
global

Commit:
switching inserting joinfiles to an int

Commit:
remove debugging use of dol_syslog

Commit:
not defined that early

Commit:
update email_template using both ID and label. Label function even handles changing the label

Commit:
joinfiles is an integer, so no need to escape

Commit:
fixing phpstan errors

JonBendtsen commented:
I don't really see any errors inphan or phpstan that seems to be from files I added.

Yes I know that there is that `html.formmail.class.php` but I just removed the entire 2. class definition in that file and made no changes to the top.

So I think it's about ready :-)

JonBendtsen commented:
I think that it might just be possible to change the ref of an object to another ref that already exists :-(

eldy committed:
Update cemailtemplate.class.php

eldy committed:
Update cemailtemplate.class.php

eldy committed:
Update cemailtemplate.class.php

Commit:
remove public function isExistingObject because it is implemented in commonobject.class.php

JonBendtsen commented:
all subendpoints of this API

<img width="990" height="407" alt="image" src="https://github.com/user-attachments/assets/81bdf6bc-cbf4-4d76-bef0-698f4e9e47eb" />

Commit:
attempt at fixing build issues

Commit:
escaping for database

Commit:
entity from getEntity is always an integer, so no need to type cast it

Commit:
escaping more SQL statements

Commit:
adding type casts to the same as in the public function create where missing

Commit:
type casting in the api file

JonBendtsen commented:
This PR was finalised during devcamp vienna 2025

Commit:
switching from old ModelMail to new cEmailTemplate

JonBendtsen commented on htdocs/core/class/cemailtemplate.class.php:
```diff
@@ -0,0 +1,582 @@
+ * old class name for Object of table llx_c_email_templates
+ * I prefer the cEmailTemplate name as it better reflects the database
+ */
+class ModelMail extends cEmailTemplate
```
@eldy  I think this PR is ready. At what point can we deprecate the old name `ModelMail` and remove this class definition?

It does not seem to be used in ANY of the files in this github repo, but maybe someone uses it in a module somewhere?

hregis committed:
FIX wrong entity filter (#35691)

* FIX wrong entity filter

* FIX travis error

atm-quentin committed:
Fix: [Import] Prevents $keyfield variable overwrite when processing hidden fields (#35693)

* FIX: prevent duplicate key fields and improve field handling in import_xlsx and import_csv modules

* FIX: remove leftover debug statement in import_xlsx module

Commit:
no need to use db->esscape on integers

Commit:
some PHPstan fixes in update function

Commit:
fixing more phpstan errors in public function update

Commit:
final PHPStan fixes?

Commit:
handling when fetch fails

Commit:
making update by id and label work

Commit:
getting the entity

Commit:
using setEntity to get entity right

Commit:
With hregis changes and a multientity hurl test

aspangaro committed:
FIX Accountancy - Print subledger balance return general balance (#35712)

* FIX Accountancy - Print subledger balance return general balance

* Update pdf_balance.modules.php

---------

Co-authored-by: Laurent Destailleur <eldy@destailleur.fr>

atm-corentin committed:
Fix : Display supplier prices for shared products in multicompany context (#35600)

* Update SQL join condition for supplier prices to respect multi-entity context.

* Update SQL join condition for product supplier prices to use correct entity.

---------

Co-authored-by: Laurent Destailleur <eldy@destailleur.fr>

eldy committed:
Update security.txt

eldy committed:
FIX Deletion of a donation. Button was disabled.

eldy committed:
Fix look and feel

eldy committed:
Debug the embeded dolistore search tool

eldy committed:
FIX prices must be HT in dolistore browser

eldy committed:
Fix the price in module browser

eldy committed:
Fix responsive

eldy committed:
Trans

eldy committed:
Fix CSS regression. The arrow was no more visible with image using font
awesome.

eldy committed:
Replace function img_pdf with img_picto

eldy committed:
Clean code

eldy committed:
Debug v23

eldy committed:
Debug v23

eldy committed:
PERF Remove distinct and join in list of agenda events

eldy committed:
NEW Add option PROJECT_CAN_ALWAYS_LINK_TO_ALL_CUSTOMERS

eldy committed:
FIX The status of customer was not synchronized with opportunity status

eldy committed:
FIX Status if thirdparty not synchronized with status WON of
opportunity.

eldy committed:
Add anchor

eldy committed:
Clean setup pages

eldy committed:
Clean setup to reduce puplicate pages

eldy committed:
Clean setup pages

eldy committed:
Clean setup pages

eldy committed:
Fix ci

eldy committed:
Debug v23

eldy committed:
FIX link to sort of target email page

eldy committed:
Clean code

eldy committed:
FIX Bad label for column title on multicurency

eldy committed:
Better message

eldy committed:
Debug v23

hregis committed:
FIX wrong field name (#35728)

frederic34 committed:
clean code in movement_list.php (#35731)

* clean code in movement_list.php

* Update externalModules.class.php

---------

Co-authored-by: Laurent Destailleur <eldy@destailleur.fr>

Hystepik committed:
Fix member doc gen (#35751)

Co-authored-by: Lucas Marcouiller <lmarcouiller@dolicloud.com>

frederic34 committed:
fix CI v22 (#35737)

* fix CI v22

* $error is already checked

frederic34 committed:
fix img_pictos (#35722)

* fix img_pictos

* fix img_pictos

* fix img_pictos

* fix img_pictos

* fix img_pictos

* fix img_pictos

* fix img_pictos

* fix img_pictos

JonBendtsen committed:
Reviving some old PR with API get, activate and disable modules (#35718)

* Reviving some old PR with API get, activate and disable modules

* expanding hurl test for setup wih some module tests

* return array, not object

* fix PHPStan on getmodules

* typecasting module functions to string

* instiating class DolibarrModule into an object so we can call functions on it

---------

Co-authored-by: Jon Bendtsen <xcodeauthor@jonb.dk>

aspangaro committed:
FIX Accountancy - Printing the subsidiary ledger returns the general ledger (#35719)

* Update listbyaccount to transfer type="sub" information

* Detail by subsidiary account

* Fix typo

aspangaro committed:
FIX Accountancy - Admin personalized report - Remove duplicate button (#35721)

github-actions[bot] committed:
PHPStan > Update baseline (#35715)

Co-authored-by: Dolibot <dolibarr-bot@users.noreply.github.com>

frederic34 committed:
clean code (#35710)

* clean code

* clean code

* clean code

* clean code

* clean code

* clean code

* clean code

* clean code

* clean code

* clean code

* clean code

* fix

* clean code

estebanthi committed:
CLOSE #35716 Supplier invoices API endpoint to mark invoice as paid or unpaid (#35717)

Supplier invoices API endpoints to mark invoice as paid or unpaid, implemented exactly the same way it is implemented for client invoices.

aspangaro committed:
NEW Accountancy - Add a protection on various payment for auxiliary account on general account not centralized (#35720)

eldy committed:
Doc

eldy committed:
Merge branch 'develop' of git@github.com:Dolibarr/dolibarr.git into
develop

eldy committed:
Fi CI

eldy committed:
Fix CI

JonBendtsen committed:
Fix: 3 modules where list.php redirected to /mymodule/myobject_card.php (#35732)

* Fix: 3 modules where list.php redirected to /mymodule/myobject_card.php

* remove an commented segment of the code which did the same kind of dolbuildpath to /mymodule/myobject_card.php which is obviously wrong, since it was a comment anyway, no risk

* remove an commented line of the code which did the same kind of dolbuildpath to /mymodule/myobject_card.php which is obviously wrong, since it was a comment anyway, no risk

---------

Co-authored-by: Jon Bendtsen <xcodeauthor@jonb.dk>

eldy committed:
Fix trans

JonBendtsen committed:
api_products purchase_prices: Adjust RestException with missing numbers (#35749)

Co-authored-by: Jon Bendtsen <xcodeauthor@jonb.dk>
Co-authored-by: Laurent Destailleur <eldy@destailleur.fr>

JonBendtsen committed:
Mass mailing list: make title a clickable link (#35746)

* Mass mailing list: make title a clickable link

* Update list.php

---------

Co-authored-by: Jon Bendtsen <xcodeauthor@jonb.dk>
Co-authored-by: Laurent Destailleur <eldy@destailleur.fr>

defrance committed:
warning error if realpath is not empty before use (#35740)

clement-jny committed:
fix: conditionally include extrafields for filtering (#35742)

Co-authored-by: Laurent Destailleur <eldy@destailleur.fr>

eldy committed:
Clean code

frederic34 committed:
clean code (#35736)

* clean code

* no element line

* no element line

* no element line

* no element line

---------

Co-authored-by: Laurent Destailleur <eldy@destailleur.fr>

defrance committed:
Modify getListOfContactTypes for expedition type (#35726)

* Modify getListOfContactTypes for expedition type

Adjust the contact type based on shipping settings.

* Update api_setup.class.php

* Update api_setup.class.php

---------

Co-authored-by: Laurent Destailleur <eldy@destailleur.fr>

eldy committed:
Fix #35754

eldy committed:
Close #35754 - Better fix in 1 POST.

JonBendtsen committed:
FIX #35318 better array check fields_label (#35750)

* FIX #35318 better array check fields_label

* Update html.formwebportal.class.php

* Update extrafields.class.php

* Update commonobject.class.php

* Update ajaxextrafield.php

---------

Co-authored-by: Jon Bendtsen <xcodeauthor@jonb.dk>
Co-authored-by: Laurent Destailleur <eldy@destailleur.fr>

JonBendtsen committed:
Api member createSubscription, not all possible RestException numbers was mentioned (#35748)

Co-authored-by: Jon Bendtsen <xcodeauthor@jonb.dk>

JonBendtsen committed:
New: Adding linked project to mass mailing card (#35723)

* New: Adding linked project to mass mailing card

* modify project on mailing card

* mailing card load fk_project pre commit fix

* only fetch projects with id>0

* fix PHPstan errors regarding mesg

* projects related to mass mailings does not have a soc or socid

* remove last trace of soc and socid

* checking if mesg is set before using it

* just loading all projects, because maybe you have a mailing list with workers for a thirdparty? think company party

* showing the project during edit

* given that action is already something else, it will never not be classify

* making PHPstan happy

* Adding project column to list of mass mailings. Search, order and click able

* pasted the wrong place

* no need to check for filtermail

* x now also clears project search

---------

Co-authored-by: Jon Bendtsen <xcodeauthor@jonb.dk>
Co-authored-by: Laurent Destailleur <eldy@destailleur.fr>

eldy committed:
Debug v23

eldy committed:
FIX The IBAN into EPC qr code must use the default bank account if not
forced

JonBendtsen commented:
being replaced with https://github.com/Dolibarr/dolibarr/pull/35853
